src/backend/gl/src/lib.rs

```diff
@@ -921,6 +875,55 @@ impl c::Device for Device {
         fence_opt.unwrap_or_else(|| self.place_fence())
     }
 
+    fn wait_fence(&mut self, fence: &handle::Fence<Self::Resources>) {
+        wait_fence(self.frame_handles.ref_fence(&fence), &self.share.context);
+    }
+
+    fn read_mapping<'a, 'b, M, T>(&'a mut self, m: &'b mut M)
```

Comment:
Please move the mapping related methods to the `Factory`. The device should be responsible for commands submission, not memory management.